# Haonan Xu

![title](../p1.jpeg)

![title](../p2.jpeg)

In [1]:
import pandas as pd 
from sklearn.multiclass import OneVsRestClassifier
from sklearn.svm import SVC
from sklearn import svm
import numpy as np
from sklearn.metrics import accuracy_score

In [2]:
!pip install imblearn

In [3]:
data = pd.read_csv("../data/Frogs_MFCCs.csv") 

In [4]:
data

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  MFCCs_ 4  MFCCs_ 5  MFCCs_ 6  MFCCs_ 7  \
0          1.0  0.152936 -0.105586  0.200722  0.317201  0.260764  0.100945   
1          1.0  0.171534 -0.098975  0.268425  0.338672  0.268353  0.060835   
2          1.0  0.152317 -0.082973  0.287128  0.276014  0.189867  0.008714   
3          1.0  0.224392  0.118985  0.329432  0.372088  0.361005  0.015501   
4          1.0  0.087817 -0.068345  0.306967  0.330923  0.249144  0.006884   
...        ...       ...       ...       ...       ...       ...       ...   
7190       1.0 -0.554504 -0.337717  0.035533  0.034511  0.443451  0.093889   
7191       1.0 -0.517273 -0.370574  0.030673  0.068097  0.402890  0.096628   
7192       1.0 -0.582557 -0.343237  0.029468  0.064179  0.385596  0.114905   
7193       1.0 -0.519497 -0.307553 -0.004922  0.072865  0.377131  0.086866   
7194       1.0 -0.508833 -0.324106  0.062068  0.078211  0.397188  0.094596   

      MFCCs_ 8  MFCCs_ 9  MFCCs_10  ...  MFCCs_17  MFCCs_18  MFCCs_19  \
0    -0.150063 -0.171128  0.124676  ... -0.108351 -0.077623 -0.009568   
1    -0.222475 -0.207693  0.170883  ... -0.090974 -0.056510 -0.035303   
2    -0.242234 -0.219153  0.232538  ... -0.050691 -0.023590 -0.066722   
3    -0.194347 -0.098181  0.270375  ... -0.136009 -0.177037 -0.130498   
4    -0.265423 -0.172700  0.266434  ... -0.048885 -0.053074 -0.088550   
...        ...       ...       ...  ...       ...       ...       ...   
7190 -0.100753  0.037087  0.081075  ...  0.069430  0.071001  0.021591   
7191 -0.116460  0.063727  0.089034  ...  0.061127  0.068978  0.017745   
7192 -0.103317  0.070370  0.081317  ...  0.082474  0.077771 -0.009688   
7193 -0.115799  0.056979  0.089316  ...  0.051796  0.069073  0.017963   
7194 -0.117672  0.058874  0.076180  ...  0.061455  0.072983 -0.003980   

      MFCCs_20  MFCCs_21  MFCCs_22           Family      Genus  \
0     0.057684  0.118680  0.014038  Leptodactylidae  Adenomera   
1     0.020140  0.082263  0.029056  Leptodactylidae  Adenomera   
2    -0.025083  0.099108  0.077162  Leptodactylidae  Adenomera   
3    -0.054766 -0.018691  0.023954  Leptodactylidae  Adenomera   
4    -0.031346  0.108610  0.079244  Leptodactylidae  Adenomera   
...        ...       ...       ...              ...        ...   
7190  0.052449 -0.021860 -0.079860          Hylidae     Scinax   
7191  0.046461 -0.015418 -0.101892          Hylidae     Scinax   
7192  0.027834 -0.000531 -0.080425          Hylidae     Scinax   
7193  0.041803 -0.027911 -0.096895          Hylidae     Scinax   
7194  0.031560 -0.029355 -0.087910          Hylidae     Scinax   

             Species  RecordID  
0     AdenomeraAndre         1  
1     AdenomeraAndre         1  
2     AdenomeraAndre         1  
3     AdenomeraAndre         1  
4     AdenomeraAndre         1  
...              ...       ...  
7190     ScinaxRuber        60  
7191     ScinaxRuber        60  
7192     ScinaxRuber        60  
7193     ScinaxRuber        60  
7194     ScinaxRuber        60  

[7195 rows x 26 columns]

# b(i) Evaluating use exact match and hamming score

### Accuracy score/ Exact match metric:
### This function calculates subset accuracy meaning the predicted set of labels should exactly match
### with the true set of labels.


### start with single label multiclass problem

In [5]:

X = data.iloc[:, 0:22]
y_1 = data.iloc[:, 22:23]
y_2 = data.iloc[:, 23:24]
y_3 = data.iloc[:, 24:25]
y_true = data.iloc[:, 22:25]
y_true = y_true.to_numpy()

In [6]:
X 

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  MFCCs_ 4  MFCCs_ 5  MFCCs_ 6  MFCCs_ 7  \
0          1.0  0.152936 -0.105586  0.200722  0.317201  0.260764  0.100945   
1          1.0  0.171534 -0.098975  0.268425  0.338672  0.268353  0.060835   
2          1.0  0.152317 -0.082973  0.287128  0.276014  0.189867  0.008714   
3          1.0  0.224392  0.118985  0.329432  0.372088  0.361005  0.015501   
4          1.0  0.087817 -0.068345  0.306967  0.330923  0.249144  0.006884   
...        ...       ...       ...       ...       ...       ...       ...   
7190       1.0 -0.554504 -0.337717  0.035533  0.034511  0.443451  0.093889   
7191       1.0 -0.517273 -0.370574  0.030673  0.068097  0.402890  0.096628   
7192       1.0 -0.582557 -0.343237  0.029468  0.064179  0.385596  0.114905   
7193       1.0 -0.519497 -0.307553 -0.004922  0.072865  0.377131  0.086866   
7194       1.0 -0.508833 -0.324106  0.062068  0.078211  0.397188  0.094596   

      MFCCs_ 8  MFCCs_ 9  MFCCs_10  ...  MFCCs_13  MFCCs_14  MFCCs_15  \
0    -0.150063 -0.171128  0.124676  ... -0.156436  0.082245  0.135752   
1    -0.222475 -0.207693  0.170883  ... -0.254341  0.022786  0.163320   
2    -0.242234 -0.219153  0.232538  ... -0.237384  0.050791  0.207338   
3    -0.194347 -0.098181  0.270375  ... -0.317084 -0.011567  0.100413   
4    -0.265423 -0.172700  0.266434  ... -0.298524  0.037439  0.219153   
...        ...       ...       ...  ...       ...       ...       ...   
7190 -0.100753  0.037087  0.081075  ... -0.145668 -0.059364  0.024206   
7191 -0.116460  0.063727  0.089034  ... -0.164675 -0.105600  0.030767   
7192 -0.103317  0.070370  0.081317  ... -0.150025 -0.078615  0.024861   
7193 -0.115799  0.056979  0.089316  ... -0.153120 -0.075320  0.022903   
7194 -0.117672  0.058874  0.076180  ... -0.150554 -0.073415  0.042517   

      MFCCs_16  MFCCs_17  MFCCs_18  MFCCs_19  MFCCs_20  MFCCs_21  MFCCs_22  
0    -0.024017 -0.108351 -0.077623 -0.009568  0.057684  0.118680  0.014038  
1     0.012022 -0.090974 -0.056510 -0.035303  0.020140  0.082263  0.029056  
2     0.083536 -0.050691 -0.023590 -0.066722 -0.025083  0.099108  0.077162  
3    -0.050224 -0.136009 -0.177037 -0.130498 -0.054766 -0.018691  0.023954  
4     0.062837 -0.048885 -0.053074 -0.088550 -0.031346  0.108610  0.079244  
...        ...       ...       ...       ...       ...       ...       ...  
7190 -0.000861  0.069430  0.071001  0.021591  0.052449 -0.021860 -0.079860  
7191  0.006457  0.061127  0.068978  0.017745  0.046461 -0.015418 -0.101892  
7192  0.008696  0.082474  0.077771 -0.009688  0.027834 -0.000531 -0.080425  
7193  0.001924  0.051796  0.069073  0.017963  0.041803 -0.027911 -0.096895  
7194  0.004158  0.061455  0.072983 -0.003980  0.031560 -0.029355 -0.087910  

[7195 rows x 22 columns]

In [7]:
y_1

Family
0     Leptodactylidae
1     Leptodactylidae
2     Leptodactylidae
3     Leptodactylidae
4     Leptodactylidae
...               ...
7190          Hylidae
7191          Hylidae
7192          Hylidae
7193          Hylidae
7194          Hylidae

[7195 rows x 1 columns]

###  Gaussian kernel = RBF kernel:  ||X1 — X2 || = Euclidean distance between X1 & X2

In [8]:
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler

from sklearn.svm import SVC
## parameters: rbf/gaussian kernel, C=, gamma = , one vs rest  
# Train a classifier for each label 
clf_1 = make_pipeline(StandardScaler(), SVC(C=0.1, kernel='rbf',gamma=0.1, decision_function_shape='ovr'))
clf_1.fit(X, y_1)


clf_2 = make_pipeline(StandardScaler(), SVC(C=0.1, kernel='rbf',gamma=0.01, decision_function_shape='ovr'))
clf_2.fit(X, y_2)


clf_3 = make_pipeline(StandardScaler(), SVC(C=0.1, kernel='rbf',gamma=0.01,decision_function_shape='ovr'))
clf_3.fit(X, y_3)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('svc', SVC(C=0.1, gamma=0.01))])

In [9]:
predict_y_1 = clf_1.predict(X)
    

In [10]:
predict_y_2 = clf_2.predict(X)

In [11]:
predict_y_3 = clf_3.predict(X)

In [12]:
predict_multi_label = np.array(list(zip(predict_y_1,predict_y_2,predict_y_3)))
predict_multi_label

array([['Leptodactylidae', 'Ameerega', 'Ameeregatrivittata'],
       ['Leptodactylidae', 'Ameerega', 'AdenomeraAndre'],
       ['Leptodactylidae', 'Adenomera', 'AdenomeraAndre'],
       ...,
       ['Hylidae', 'Ameerega', 'Ameeregatrivittata'],
       ['Hylidae', 'Ameerega', 'Ameeregatrivittata'],
       ['Hylidae', 'Ameerega', 'Ameeregatrivittata']], dtype='<U22')

In [13]:
from sklearn.metrics import accuracy_score
y_pred = predict_multi_label
length = predict_multi_label.shape[0]

num_row = length
num_match=0

for i in range(length):
    if np.array_equal(y_true[i], y_pred[i]):
        num_match=num_match+1

print(num_match/num_row)

0.9188325225851286


### After trying different extreme values above for C and gamma(lamda in lecture notes),   I found that
### the accuracy of a svm will not be below 70% for C(0.004,0.1), gamma(0.001.0.1)
### Use them in as the range during cross validation 

In [14]:
# for label 1
from sklearn.model_selection import GridSearchCV
parameters = {'C':[0.004,0.01,0.04,0.1],'gamma':[0.001,0.005,0.01,0.05,0.1]}
svc = svm.SVC(kernel='rbf')
clf = GridSearchCV(svc, parameters, cv=10)
clf.fit(X, y_1)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

GridSearchCV(cv=10, estimator=SVC(),
             param_grid={'C': [0.004, 0.01, 0.04, 0.1],
                         'gamma': [0.001, 0.005, 0.01, 0.05, 0.1]})

In [15]:
clf.best_params_

{'C': 0.1, 'gamma': 0.1}

In [16]:
clf_1 = make_pipeline(StandardScaler(), SVC(C=0.1, kernel='rbf',gamma=0.1, decision_function_shape='ovr'))
clf_1.fit(X, y_1)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('svc', SVC(C=0.1, gamma=0.1))])

In [17]:
y_pred_1 = clf_1.predict(X)

In [18]:
#for label 2
parameters = {'C':[0.004,0.01,0.04,0.1],'gamma':[0.001,0.005,0.01,0.05,0.1]}
svc = svm.SVC(kernel='rbf')
clf = GridSearchCV(svc, parameters, cv=10)
clf.fit(X, y_2)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

GridSearchCV(cv=10, estimator=SVC(),
             param_grid={'C': [0.004, 0.01, 0.04, 0.1],
                         'gamma': [0.001, 0.005, 0.01, 0.05, 0.1]})

In [19]:
clf.best_params_

{'C': 0.1, 'gamma': 0.1}

In [20]:
clf_2 = make_pipeline(StandardScaler(), SVC(C=0.1, kernel='rbf',gamma=0.1, decision_function_shape='ovr'))
clf_2.fit(X, y_2)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('svc', SVC(C=0.1, gamma=0.1))])

In [21]:
y_pred_2 = clf_2.predict(X)

In [22]:
# label 3
parameters = {'C':[0.004,0.01,0.04,0.1],'gamma':[0.001,0.005,0.01,0.05,0.1]}
svc = svm.SVC(kernel='rbf')
clf = GridSearchCV(svc, parameters, cv=10)
clf.fit(X, y_3)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

GridSearchCV(cv=10, estimator=SVC(),
             param_grid={'C': [0.004, 0.01, 0.04, 0.1],
                         'gamma': [0.001, 0.005, 0.01, 0.05, 0.1]})

In [23]:
clf.best_params_

{'C': 0.1, 'gamma': 0.1}

In [24]:
clf_3 = make_pipeline(StandardScaler(), SVC(C=0.1, kernel='rbf',gamma=0.1,decision_function_shape='ovr'))
clf_3.fit(X, y_3)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('svc', SVC(C=0.1, gamma=0.1))])

In [25]:
y_pred_3 = clf_3.predict(X)

In [26]:
y_pred = np.array(list(zip(y_pred_1,y_pred_2,y_pred_3)))
y_pred

array([['Leptodactylidae', 'Adenomera', 'AdenomeraAndre'],
       ['Leptodactylidae', 'Adenomera', 'AdenomeraAndre'],
       ['Leptodactylidae', 'Adenomera', 'AdenomeraAndre'],
       ...,
       ['Hylidae', 'Scinax', 'ScinaxRuber'],
       ['Hylidae', 'Scinax', 'ScinaxRuber'],
       ['Hylidae', 'Scinax', 'ScinaxRuber']], dtype='<U22')

In [27]:
# EXACT MATCH
def accuracy(y_true, y_pred):
    length = y_pred.shape[0]

    num_row = length
    num_match=0

    for i in range(length):
        if np.array_equal(y_true[i], y_pred[i]):
            num_match=num_match+1
    return num_match, num_match/num_row

_,acc = accuracy(y_true,y_pred)
print('accuracy score for exact match is: ', acc)

accuracy score for exact match is:  0.9577484364141765


## accuracy score for exact match is:  0.9577484364141765

In [28]:
# HAMMING SCORE
y_true_1 = y_1.to_numpy().ravel()
num_1,_ = accuracy(y_true_1, y_pred_1)
num_1

7070

In [29]:
y_true_2 = y_2.to_numpy().ravel()
num_2,_ = accuracy(y_true_2, y_pred_2)
num_2

6994

In [30]:
y_true_3 = y_3.to_numpy().ravel()
num_3,_ = accuracy(y_true_3, y_pred_3)
num_3

6957

In [31]:
print('hamming score is: ', (num_1+num_2+num_3)/(y_pred_1.shape[0]*3))

hamming score is:  0.9738707435719249


## hamming score is:  0.9738707435719249

# b(ii) Repeat with L1-penalized SVM

In [32]:
from sklearn.svm import LinearSVC

clf_1 = make_pipeline(StandardScaler(), LinearSVC(C=0.001, penalty='l1',dual=False))
clf_1.fit(X, y_1)


clf_2 = make_pipeline(StandardScaler(), LinearSVC(C=0.001, penalty='l1',dual=False))
clf_2.fit(X, y_2)


clf_3 = make_pipeline(StandardScaler(), LinearSVC(C=0.001, penalty='l1',dual=False))
clf_3.fit(X, y_3)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('linearsvc', LinearSVC(C=0.001, dual=False, penalty='l1'))])

In [33]:
predict_y_1 = clf_1.predict(X)
predict_y_2 = clf_2.predict(X)
predict_y_3 = clf_3.predict(X)

In [34]:
predict_multi_label = np.array(list(zip(predict_y_1,predict_y_2,predict_y_3)))
predict_multi_label

array([['Dendrobatidae', 'Ameerega', 'AdenomeraAndre'],
       ['Hylidae', 'Ameerega', 'AdenomeraAndre'],
       ['Leptodactylidae', 'Adenomera', 'AdenomeraAndre'],
       ...,
       ['Hylidae', 'Hypsiboas', 'HypsiboasCordobae'],
       ['Hylidae', 'Hypsiboas', 'HypsiboasCordobae'],
       ['Hylidae', 'Hypsiboas', 'HypsiboasCordobae']], dtype='<U22')

In [35]:
y_pred = predict_multi_label
length = predict_multi_label.shape[0]

num_row = length
num_match=0

for i in range(length):
    if np.array_equal(y_true[i], y_pred[i]):
        num_match=num_match+1

print(num_match/num_row)

0.7797081306462822


## The C interval is [0.001, 1] this time

In [36]:
parameters = {'C':[0.001,0.01,0.1,1]}
svc = LinearSVC(penalty='l1',dual=False)
clf = GridSearchCV(svc, parameters, cv=10)
clf.fit(X, y_1)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of 

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haona

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haona

GridSearchCV(cv=10, estimator=LinearSVC(dual=False, penalty='l1'),
             param_grid={'C': [0.001, 0.01, 0.1, 1]})

In [37]:
clf.best_params_

{'C': 1}

In [38]:
clf_1 = make_pipeline(StandardScaler(), LinearSVC(C=1, penalty='l1',dual=False))
clf_1.fit(X, y_1)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('linearsvc', LinearSVC(C=1, dual=False, penalty='l1'))])

In [39]:
y_pred_1 = clf_1.predict(X)

In [40]:
parameters = {'C':[0.001,0.01,0.1,1]}
svc = LinearSVC(penalty='l1',dual=False)
clf = GridSearchCV(svc, parameters, cv=10)
clf.fit(X, y_2)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haona

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haona

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haona

GridSearchCV(cv=10, estimator=LinearSVC(dual=False, penalty='l1'),
             param_grid={'C': [0.001, 0.01, 0.1, 1]})

In [41]:
clf.best_params_

{'C': 1}

In [42]:
clf_2 = make_pipeline(StandardScaler(), LinearSVC(C=1, penalty='l1',dual=False))
clf_2.fit(X, y_2)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('linearsvc', LinearSVC(C=1, dual=False, penalty='l1'))])

In [43]:
y_pred_2 = clf_2.predict(X)

In [44]:
parameters = {'C':[0.001,0.01,0.1,1]}
svc = LinearSVC(penalty='l1',dual=False)
clf = GridSearchCV(svc, parameters, cv=10)
clf.fit(X, y_3)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haona

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haona

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haona

GridSearchCV(cv=10, estimator=LinearSVC(dual=False, penalty='l1'),
             param_grid={'C': [0.001, 0.01, 0.1, 1]})

In [45]:
clf.best_params_

{'C': 1}

In [46]:
clf_3 = make_pipeline(StandardScaler(), LinearSVC(C=1, penalty='l1',dual=False))
clf_3.fit(X, y_3)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('linearsvc', LinearSVC(C=1, dual=False, penalty='l1'))])

In [47]:
y_pred_3 = clf_3.predict(X)

In [48]:
y_pred = np.array(list(zip(y_pred_1,y_pred_2,y_pred_3)))
y_pred

array([['Dendrobatidae', 'Ameerega', 'Ameeregatrivittata'],
       ['Dendrobatidae', 'Ameerega', 'Ameeregatrivittata'],
       ['Dendrobatidae', 'Adenomera', 'AdenomeraAndre'],
       ...,
       ['Hylidae', 'Scinax', 'ScinaxRuber'],
       ['Hylidae', 'Scinax', 'ScinaxRuber'],
       ['Hylidae', 'Scinax', 'ScinaxRuber']], dtype='<U22')

In [49]:
# EXACT MATCH
def accuracy(y_true, y_pred):
    length = y_pred.shape[0]

    num_row = length
    num_match=0

    for i in range(length):
        if np.array_equal(y_true[i], y_pred[i]):
            num_match=num_match+1
    return num_match, num_match/num_row

_,acc = accuracy(y_true,y_pred)
print('accuracy score for exact match is: ', acc)

accuracy score for exact match is:  0.9192494788047255


## accuracy score for exact match on L1-penalized SVM is:  0.9193884642112579

In [50]:
#Hamming
y_true_1 = y_1.to_numpy().ravel()
num_1,_ = accuracy(y_true_1, y_pred_1)
num_1

6748

In [51]:
y_true_2 = y_2.to_numpy().ravel()
num_2,_ = accuracy(y_true_2, y_pred_2)
num_2

6854

In [52]:
y_true_3 = y_3.to_numpy().ravel()
num_3,_ = accuracy(y_true_3, y_pred_3)
num_3

6917

In [53]:
print('hamming score is: ', (num_1+num_2+num_3)/(y_pred_1.shape[0]*3))

hamming score is:  0.9506138522121844


## hamming score for L1-penalized SVM is:  0.9507528376187167

# b(iv) with SMOTE data

In [54]:
from imblearn.over_sampling import SMOTE 
sm = SMOTE(random_state=2)
X_1_res, y_1_res = sm.fit_sample(X, y_1)

In [55]:
X_2_res, y_2_res = sm.fit_sample(X, y_2)

In [56]:
X_3_res, y_3_res = sm.fit_sample(X, y_3)

In [57]:
clf_1 = make_pipeline(StandardScaler(), LinearSVC(C=1000, penalty='l1',dual=False))
clf_1.fit(X_1_res, y_1_res)


clf_2 = make_pipeline(StandardScaler(), LinearSVC(C=1000, penalty='l1',dual=False))
clf_2.fit(X_2_res, y_2_res)


clf_3 = make_pipeline(StandardScaler(), LinearSVC(C=1000, penalty='l1',dual=False))
clf_3.fit(X_3_res, y_3_res)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haona

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('linearsvc', LinearSVC(C=1000, dual=False, penalty='l1'))])

In [58]:
predict_y_1 = clf_1.predict(X)
predict_y_2 = clf_2.predict(X)
predict_y_3 = clf_3.predict(X)

In [59]:
predict_multi_label = np.array(list(zip(predict_y_1,predict_y_2,predict_y_3)))
predict_multi_label

array([['Dendrobatidae', 'Ameerega', 'Ameeregatrivittata'],
       ['Dendrobatidae', 'Ameerega', 'Ameeregatrivittata'],
       ['Dendrobatidae', 'Dendropsophus', 'Ameeregatrivittata'],
       ...,
       ['Hylidae', 'Scinax', 'ScinaxRuber'],
       ['Hylidae', 'Scinax', 'ScinaxRuber'],
       ['Hylidae', 'Scinax', 'ScinaxRuber']], dtype='<U22')

In [60]:
y_pred = predict_multi_label
length = predict_multi_label.shape[0]

num_row = length
num_match=0

for i in range(length):
    if np.array_equal(y_true[i], y_pred[i]):
        num_match=num_match+1

print(num_match/num_row)

0.8643502432244614


## C: [0.001, 100]

In [61]:
parameters = {'C':[0.001,0.01,0.1,1,10,100]}
svc = LinearSVC(penalty='l1',dual=False)
clf = GridSearchCV(svc, parameters, cv=10)
clf.fit(X_1_res, y_1_res)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haona

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haona

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haona

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haona

GridSearchCV(cv=10, estimator=LinearSVC(dual=False, penalty='l1'),
             param_grid={'C': [0.001, 0.01, 0.1, 1, 10, 100]})

In [62]:
clf.best_params_

{'C': 100}

In [63]:
clf_1 = make_pipeline(StandardScaler(), LinearSVC(C=100, penalty='l1',dual=False))
clf_1.fit(X_1_res, y_1_res)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('linearsvc', LinearSVC(C=100, dual=False, penalty='l1'))])

In [64]:
y_pred_1 = clf_1.predict(X)

In [65]:
#parameters = {'C':[0.1,1,10,100]}
#svc = LinearSVC(penalty='l1',dual=False, max_iter=10000)
#clf = GridSearchCV(svc, parameters, cv=10)
#clf.fit(X_2_res, y_2_res)

In [66]:
#clf.best_params_

### To save time, I don't re-run the above block, but the result is C=100

In [67]:
clf_2 = make_pipeline(StandardScaler(), LinearSVC(C=100, penalty='l1',dual=False))
clf_2.fit(X_2_res, y_2_res)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('linearsvc', LinearSVC(C=100, dual=False, penalty='l1'))])

In [68]:
y_pred_2 = clf_2.predict(X)

In [69]:
#parameters = {'C':[0.1,1,10,100]}
#svc = LinearSVC(penalty='l1',dual=False, max_iter=10000)
#clf = GridSearchCV(svc, parameters, cv=10)
#clf.fit(X_3_res, y_3_res)

In [70]:
#clf.best_params_

In [71]:
clf_3 = make_pipeline(StandardScaler(), LinearSVC(C=100, penalty='l1',dual=False))
clf_3.fit(X_3_res, y_3_res)

/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/Users/haonanxu/opt/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:976: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('linearsvc', LinearSVC(C=100, dual=False, penalty='l1'))])

In [72]:
y_pred_3 = clf_3.predict(X)

In [73]:
y_pred = np.array(list(zip(y_pred_1,y_pred_2,y_pred_3)))
y_pred

array([['Dendrobatidae', 'Ameerega', 'Ameeregatrivittata'],
       ['Dendrobatidae', 'Ameerega', 'Ameeregatrivittata'],
       ['Dendrobatidae', 'Dendropsophus', 'Ameeregatrivittata'],
       ...,
       ['Hylidae', 'Scinax', 'ScinaxRuber'],
       ['Hylidae', 'Scinax', 'ScinaxRuber'],
       ['Hylidae', 'Scinax', 'ScinaxRuber']], dtype='<U22')

In [74]:
# EXACT MATCH
def accuracy(y_true, y_pred):
    length = y_pred.shape[0]

    num_row = length
    num_match=0

    for i in range(length):
        if np.array_equal(y_true[i], y_pred[i]):
            num_match=num_match+1
    return num_match, num_match/num_row

_,acc = accuracy(y_true,y_pred)
print('accuracy score for exact match is: ', acc)

accuracy score for exact match is:  0.8644892286309938


### The L1-penalized SVM model with data after SMOTE: accuracy score: 0.8643502432244614

In [75]:
#Hamming
y_true_1 = y_1.to_numpy().ravel()
num_1,_ = accuracy(y_true_1, y_pred_1)
num_1

6614

In [76]:
y_true_2 = y_2.to_numpy().ravel()
num_2,_ = accuracy(y_true_2, y_pred_2)
num_2

6596

In [77]:
y_true_3 = y_3.to_numpy().ravel()
num_3,_ = accuracy(y_true_3, y_pred_3)
num_3

6921

In [78]:
print('hamming score is: ', (num_1+num_2+num_3)/(y_pred_1.shape[0]*3))

hamming score is:  0.9326384063006717


### Hamming score is 0.9325920778318276

### The L1-penalized classifiers did not perform better than SVM classifiers with Gaussian kernel

## 2(a)

In [79]:
y_true

array([['Leptodactylidae', 'Adenomera', 'AdenomeraAndre'],
       ['Leptodactylidae', 'Adenomera', 'AdenomeraAndre'],
       ['Leptodactylidae', 'Adenomera', 'AdenomeraAndre'],
       ...,
       ['Hylidae', 'Scinax', 'ScinaxRuber'],
       ['Hylidae', 'Scinax', 'ScinaxRuber'],
       ['Hylidae', 'Scinax', 'ScinaxRuber']], dtype=object)

In [80]:
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
import random
r=random.randint(0,100)
#kmeans = KMeans(n_clusters=2, random_state=r).fit(X)

In [81]:
r

72

In [82]:
silhouette = []
for k in range(2,51):# silhouette seems not working for k=1
    kmeans = KMeans(n_clusters=k, random_state=r, max_iter=10000).fit(X)
    y_pred = kmeans.predict(X)
    tmp = silhouette_score(X, y_pred)
    silhouette.append(tmp)

In [83]:
max_score=0
max_pos=-1
for i in range(len(silhouette)):
    if(silhouette[i]>max_score):
        max_score = silhouette[i]
        max_pos=i+2
best_k=max_pos
print(max_pos,max_score)

4 0.3787509343305295


### k=4 is the best k in k-mean clustering chosen by silhouette method

## 2(b)

In [84]:
kmeans = KMeans(n_clusters=best_k, random_state=r).fit(X)

In [85]:
y_pred = kmeans.predict(X)

In [86]:
y_true = data.iloc[:, 22:25]

In [87]:
y_true['pred']=y_pred
y_true = y_true.astype(str)
y_true

Family      Genus         Species pred
0     Leptodactylidae  Adenomera  AdenomeraAndre    0
1     Leptodactylidae  Adenomera  AdenomeraAndre    0
2     Leptodactylidae  Adenomera  AdenomeraAndre    0
3     Leptodactylidae  Adenomera  AdenomeraAndre    0
4     Leptodactylidae  Adenomera  AdenomeraAndre    0
...               ...        ...             ...  ...
7190          Hylidae     Scinax     ScinaxRuber    0
7191          Hylidae     Scinax     ScinaxRuber    0
7192          Hylidae     Scinax     ScinaxRuber    0
7193          Hylidae     Scinax     ScinaxRuber    0
7194          Hylidae     Scinax     ScinaxRuber    0

[7195 rows x 4 columns]

In [88]:
(y_true['Family']+' '+y_true['pred']).value_counts()

Leptodactylidae 1    3467
Hylidae 2            1245
Leptodactylidae 2     629
Hylidae 3             590
Dendrobatidae 0       500
Leptodactylidae 0     302
Hylidae 0             229
Hylidae 1             101
Bufonidae 2            66
Dendrobatidae 2        42
Leptodactylidae 3      22
Bufonidae 3             2
dtype: int64

In [89]:
map_fam = y_true.groupby(by=['pred'])['Family'].agg(lambda x:x.value_counts().index[0])
map_fam

pred
0      Dendrobatidae
1    Leptodactylidae
2            Hylidae
3            Hylidae
Name: Family, dtype: object

In [90]:
(y_true['Genus'] + ' ' + y_true['pred']).value_counts()

Adenomera 1        3466
Hypsiboas 2        1038
Hypsiboas 3         542
Ameerega 0          500
Adenomera 2         376
Adenomera 0         296
Leptodactylus 2     253
Dendropsophus 0     189
Scinax 2             96
Dendropsophus 1      86
Osteocephalus 2      76
Rhinella 2           66
Ameerega 2           42
Scinax 0             39
Osteocephalus 3      38
Dendropsophus 2      35
Adenomera 3          12
Hypsiboas 1          12
Scinax 3             10
Leptodactylus 3      10
Leptodactylus 0       6
Scinax 1              3
Rhinella 3            2
Leptodactylus 1       1
Hypsiboas 0           1
dtype: int64

In [91]:
map_gen = y_true.groupby(by=['pred'])['Genus'].agg(lambda x:x.value_counts().index[0])
map_gen

pred
0     Ameerega
1    Adenomera
2    Hypsiboas
3    Hypsiboas
Name: Genus, dtype: object

In [92]:
(y_true['Species'] + ' ' + y_true['pred']).value_counts()

AdenomeraHylaedactylus 1    3466
HypsiboasCordobae 2         1018
Ameeregatrivittata 0         500
HypsiboasCinerascens 3       452
AdenomeraAndre 2             370
AdenomeraAndre 0             296
LeptodactylusFuscus 2        253
HylaMinuta 0                 189
ScinaxRuber 2                 96
HypsiboasCordobae 3           90
HylaMinuta 1                  86
OsteocephalusOophagus 2       76
Rhinellagranulosa 2           66
Ameeregatrivittata 2          42
ScinaxRuber 0                 39
OsteocephalusOophagus 3       38
HylaMinuta 2                  35
HypsiboasCinerascens 2        20
HypsiboasCordobae 1           12
LeptodactylusFuscus 3         10
ScinaxRuber 3                 10
AdenomeraAndre 3               6
LeptodactylusFuscus 0          6
AdenomeraHylaedactylus 3       6
AdenomeraHylaedactylus 2       6
ScinaxRuber 1                  3
Rhinellagranulosa 3            2
HypsiboasCordobae 0            1
LeptodactylusFuscus 1          1
dtype: int64

In [93]:
map_spe = y_true.groupby(by=['pred'])['Species'].agg(lambda x:x.value_counts().index[0])
map_spe

pred
0        Ameeregatrivittata
1    AdenomeraHylaedactylus
2         HypsiboasCordobae
3      HypsiboasCinerascens
Name: Species, dtype: object

## 2(c)

In [94]:
y_true['pred_fam']=""
y_true['pred_gen']=""
y_true['pred_spe']=""

y_true['pred_fam']=y_true['pred'].map(map_fam)
y_true['pred_gen']=y_true['pred'].map(map_gen)
y_true['pred_spe']=y_true['pred'].map(map_spe)

In [95]:
y_true

Family      Genus         Species pred       pred_fam  \
0     Leptodactylidae  Adenomera  AdenomeraAndre    0  Dendrobatidae   
1     Leptodactylidae  Adenomera  AdenomeraAndre    0  Dendrobatidae   
2     Leptodactylidae  Adenomera  AdenomeraAndre    0  Dendrobatidae   
3     Leptodactylidae  Adenomera  AdenomeraAndre    0  Dendrobatidae   
4     Leptodactylidae  Adenomera  AdenomeraAndre    0  Dendrobatidae   
...               ...        ...             ...  ...            ...   
7190          Hylidae     Scinax     ScinaxRuber    0  Dendrobatidae   
7191          Hylidae     Scinax     ScinaxRuber    0  Dendrobatidae   
7192          Hylidae     Scinax     ScinaxRuber    0  Dendrobatidae   
7193          Hylidae     Scinax     ScinaxRuber    0  Dendrobatidae   
7194          Hylidae     Scinax     ScinaxRuber    0  Dendrobatidae   

      pred_gen            pred_spe  
0     Ameerega  Ameeregatrivittata  
1     Ameerega  Ameeregatrivittata  
2     Ameerega  Ameeregatrivittata  
3     Ameerega  Ameeregatrivittata  
4     Ameerega  Ameeregatrivittata  
...        ...                 ...  
7190  Ameerega  Ameeregatrivittata  
7191  Ameerega  Ameeregatrivittata  
7192  Ameerega  Ameeregatrivittata  
7193  Ameerega  Ameeregatrivittata  
7194  Ameerega  Ameeregatrivittata  

[7195 rows x 7 columns]

In [96]:
# Hamming distance
fam = y_true['Family'].to_numpy()
fam_pred = y_true['pred_fam']
gen = y_true['Genus'].to_numpy()
gen_pred = y_true['pred_gen']
spe = y_true['Species'].to_numpy()
spe_pred = y_true['pred_spe']
dist1=np.sum(np.not_equal(fam, fam_pred))
dist2=np.sum(np.not_equal(gen, gen_pred))
dist3=np.sum(np.not_equal(spe, spe_pred))
hamming_dist = (dist1+dist2+dist3)/(len(fam)*3)
print('Hamming distance is: ', hamming_dist)

Hamming distance is:  0.2224229789205467


In [97]:
# Hamming score
score1=np.sum(np.equal(fam, fam_pred))
score2=np.sum(np.equal(gen, gen_pred))
score3=np.sum(np.equal(spe, spe_pred))
hamming_score = (score1+score2+score3)/(len(fam)*3)
print('Hamming score is: ', hamming_score)

Hamming score is:  0.7775770210794534


In [98]:
from sklearn.metrics import hamming_loss
print('Hamming loss is ', 1-hamming_score)

Hamming loss is  0.22242297892054663


In [99]:
import random
distances=[]
scores=[]
loss=[]
for i in range(50):
    silhouette=[]
    
    r=random.randint(0,100)
    for k in range(2,51):# silhouette seems not working for k=1
        kmeans = KMeans(n_clusters=k, random_state=r).fit(X)
        y_pred = kmeans.predict(X)
        tmp = silhouette_score(X, y_pred)
        silhouette.append(tmp)
    
    max_score=0
    max_pos=-1
    for i in range(len(silhouette)):
        if(silhouette[i]>max_score):
            max_score = silhouette[i]
            max_pos=i+2
    best_k=max_pos
            
    print("best k: ", best_k)
    
    kmeans = KMeans(n_clusters=best_k, random_state=r).fit(X)
    y_pred = kmeans.predict(X)
    y_true = data.iloc[:, 22:25]
    y_true['pred']=y_pred
    y_true = y_true.astype(str)
    
    map_fam = y_true.groupby(by=['pred'])['Family'].agg(lambda x:x.value_counts().index[0])
    map_gen = y_true.groupby(by=['pred'])['Genus'].agg(lambda x:x.value_counts().index[0])
    map_spe = y_true.groupby(by=['pred'])['Species'].agg(lambda x:x.value_counts().index[0])
    
    #print('mapping to family: ',map_fam)
    
    y_true['pred_fam']=""
    y_true['pred_gen']=""
    y_true['pred_spe']=""
    y_true['pred_fam']=y_true['pred'].map(map_fam)
    y_true['pred_gen']=y_true['pred'].map(map_gen)
    y_true['pred_spe']=y_true['pred'].map(map_spe)
    
    # Hamming distance
    fam = y_true['Family'].to_numpy()
    fam_pred = y_true['pred_fam'].to_numpy()
    gen = y_true['Genus'].to_numpy()
    gen_pred = y_true['pred_gen'].to_numpy()
    spe = y_true['Species'].to_numpy()
    spe_pred = y_true['pred_spe'].to_numpy()
    dist1=np.sum(np.not_equal(fam, fam_pred))
    dist2=np.sum(np.not_equal(gen, gen_pred))
    dist3=np.sum(np.not_equal(spe, spe_pred))
    hamming_dist = (dist1+dist2+dist3)/(len(fam)*3)
    distances.append(hamming_dist)
    
    # Hamming score
    score1=np.sum(np.equal(fam, fam_pred))
    score2=np.sum(np.equal(gen, gen_pred))
    score3=np.sum(np.equal(spe, spe_pred))
    hamming_score = (score1+score2+score3)/(len(fam)*3)
    scores.append(hamming_score)
    
    
    loss.append(1-hamming_score)
    
    print("done "+str(i+1)+" iters")
    print('hamming score is: ',hamming_score)

KeyboardInterrupt: 